### DISCUSSION TITLE
how to map OneToOne with soft foreign key?

### DISCUSSION THREAD

**Discussion by trim21** [mikro-orm/mikro-orm#6082] [General]
I have a 1:1 relation `Item{id: int primary key not null}` and `ItemExtraInfo{item_id: int primary key not null}`, and `item_extra_info.item_id` or `item.id` is not foreign keys but only normal primary keys.

```sql
create table item (
    item_id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT
);

create table item_extra_info 
    extra_item_id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    item_name     text
)
```

I tried many options but I query builder generate invalid sql:

```
... inner join `item_extra_info` as `e` on `t`.`extra_extra_item_id` = `e`.`extra_item_id` ...
```

and there is no `extra_extra_item_id` on `ItemExtraInfo`

I also tried `joinColumn="item_id"` or `joinColumn="id"` and non of them work, joinColumn can control column used on `Item` but it can't  be `item_id` becasue orm think it's "duplicated" fields with `id`

full example:
```ts
import { defineConfig, MikroORM } from '@mikro-orm/sqlite'; // or any other driver package
import { TsMorphMetadataProvider } from '@mikro-orm/reflection';
import { Entity, OneToOne, PrimaryKey, Property } from '@mikro-orm/core';


@Entity({ tableName: 'item' })
export class Item {
  @PrimaryKey({ type: 'mediumint', name: 'item_id', autoincrement: true })
  id!: number;

  @OneToOne()
  extra!: ItemExtraInfo;
}

@Entity({ tableName: 'item_extra_info' })
export class ItemExtraInfo {
  @PrimaryKey({ type: 'int', name: 'extra_item_id', autoincrement: true })
  item_id!: number;

  @Property({ type: 'text', name: 'item_name' })
  name!: string;
}


export const orm = await MikroORM.init(defineConfig({
  dbName: ':memory:',
  metadataProvider: TsMorphMetadataProvider,
  allowGlobalContext: true,
  entities: [Item, ItemExtraInfo],
}));


const conn = await orm.connect();
await conn.execute(`
    create table item
    (
        item_id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT
    );
`);

await conn.execute(`
    create table item_extra_info
    (
        extra_item_id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
        item_name     text
    )
`);

const r = await orm.em.getRepository(Item).createQueryBuilder('t').select('*').innerJoin('t.extra', 'e', { 'e.item_id': 't.id' }).where('t.id = ?', [1]).getSingleResult();

console.log(r);
```

**Comment by B4nan** [comment]
Your example defines an owning side on the item table, that's the opposite of the schema example you provided - the owning side is where the FK value is stored (it's irrelevant you don't have a FK constraint defined in the schema, this is about value mapping), so the item_extra_info table.

Try this:

```ts
@Entity()
export class Item {
  @PrimaryKey({ name: 'item_id' })
  id!: number;

  @OneToOne(() => ItemExtraInfo, item => item.ItemExtraInfo) // inverse side
  extra!: ItemExtraInfo;
}

@Entity()
export class ItemExtraInfo {
  @OneToOne({ entity: () => Item, name: 'extra_item_id', primary: true }) // owning side
  item!: Item;

  @Property({ type: 'text', name: 'item_name' })
  name!: string;
}
```

Note that it doesn't make much sense to have both of those defined as autoincrement or different type (int vs mediumint). Also note that you don't have to provide any join condition, its always inferred based on the relation.

**Reply by trim21** [reply]
thanks for your answer

**Reply by trim21** [reply]
so there is no way to define `extra_item_id` field on `ItemExtraInfo` if I want to use relation?

**Reply by B4nan** [reply]
my example is doing that, you can access the value via `item.id`, no need to populate the relation to do that

**Reply by trim21** [reply]
is it still exists if I didn't do any join? for example, `repo.qb().select('*').where(...).execite()`

**Reply by trim21** [reply]
emmm, if I didn't do any join, then `extra.item` is id of item and it's actually a number in runtime instead of `Item`, which is a type error in typescript

**Reply by B4nan** [reply]
no, its always `Item` instance on runtime with the example i provided above

**Reply by trim21** [reply]
OK, I think I asked wrong question. I want to ask is it possible to fetch `ItemExtraInfo` without fetching `Item`?

**Reply by B4nan** [reply]
yes you can, sounds like you havent really tried my example

**Reply by trim21** [reply]
Oh, sorry. I tried but `Item` doesn't have other field expect `id` and `extra`, so I think it fetch whole `Item` again. I add a new field and I can see it works as expected
